### PR TITLE
Refactor error reporting of constants

### PR DESCRIPTION
**Pull request by oli-obk**
cc @eddyb 

This PR should not change any behaviour. It solely simplifies the internal handling of the errors

### CODE REVIEW

File: src/librustc/mir/interpret/error.rs [hunk 1]
```diff
@@ -150,6 +150,9 @@ pub enum EvalErrorKind<'tcx, O> {
     UnimplementedTraitSelection,
     /// Abort in case type errors are reached
     TypeckError,
+    /// Resolution can fail if we are in a too generic context
+    ResolutionFailed,
```
Comment (eddyb) [inline src/librustc/mir/interpret/error.rs]: I'd rather call this `TooGeneric` or something more obvious.

File: src/librustc/middle/const_val.rs [hunk 2]
```diff
@@ -33,16 +33,7 @@ pub enum ConstVal<'tcx> {
...
-}
-
-#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
-pub enum ErrKind<'tcx> {
-
-    CouldNotResolve,
-    TypeckError,
-    CheckMatchError,
-    Miri(::mir::interpret::EvalError<'tcx>, Vec<FrameInfo>),
+    pub data: Lrc<(::mir::interpret::EvalError<'tcx>, Vec<FrameInfo>)>,
```
Comment (eddyb) [inline src/librustc/middle/const_val.rs]: Can this struct be moved to `mir::interpret`? Also, maybe put the `Lrc` around the struct, that way you can name the fields that are currently in a tuple.

File: src/librustc/traits/fulfill.rs [hunk 3]
```diff
@@ -508,10 +509,11 @@ impl<'a, 'b, 'gcx, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'gcx,
                                             CodeSelectionError(ConstEvalFailure(err)))
                                     }
                                 } else {
+                                    let err = EvalErrorKind::ResolutionFailed.into();
                                     ProcessResult::Error(
                                         CodeSelectionError(ConstEvalFailure(ConstEvalErr {
                                             span: obligation.cause.span,
-                                            kind: ErrKind::CouldNotResolve.into(),
+                                            data: (err, Vec::new()).into(),
```
Comment (eddyb) [inline src/librustc/traits/fulfill.rs]: nit: `vec![]`

**Comment by oli-obk** [comment]
Addressed all review comments

File: src/test/ui/const-len-underflow-separate-spans.stderr [hunk 4]
```diff
@@ -6,14 +6,23 @@ LL | const LEN: usize = ONE - TWO;
...
+   |                    --------- attempt to subtract with overflow
+...
+LL |     let a: [i8; LEN] = unimplemented!();
+   |                 ^^^
+
+error: this expression will panic at runtime
+  --> $DIR/const-len-underflow-separate-spans.rs:21:17
+   |
+LL |     let a: [i8; LEN] = unimplemented!();
+   |                 ^^^ referenced constant has errors
```
Comment (eddyb) [inline src/test/ui/const-len-underflow-separate-spans.stderr]: This seems weird, since the constant is *not* in an `Operand::Const`, but rather in a type.

Comment (oli-obk) [reply src/test/ui/const-len-underflow-separate-spans.stderr]: This note means to say that `LEN` has errors, which is a constant reference right where the span points.

File: src/librustc/mir/interpret/value.rs [hunk 5]
```diff
@@ -2,13 +2,19 @@
...
+use ty::subst::Substs;
+use hir::def_id::DefId;
 
 use super::{EvalResult, Pointer, PointerArithmetic, Allocation};
 
 /// Represents a constant value in Rust. ByVal and ScalarPair are optimizations which
 /// matches Value's optimizations for easy conversions between these two types
-#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, RustcEncodable, RustcDecodable, Hash)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, RustcEncodable, RustcDecodable, Hash)]
 pub enum ConstValue<'tcx> {
```
Comment (oli-obk) [inline src/librustc/mir/interpret/value.rs]: cc @Zoxc

File: src/test/ui/const-len-underflow-separate-spans.stderr (hunk 4 above)
Comment (eddyb) [reply src/test/ui/const-len-underflow-separate-spans.stderr]: This is an "error: this expression will panic at runtime" inside non-runtime code though?

Comment (oli-obk) [reply src/test/ui/const-len-underflow-separate-spans.stderr]: Oh... Looks like const prop is running on array lengths?! I wonder how that happened.

Comment (oli-obk) [reply src/test/ui/const-len-underflow-separate-spans.stderr]: Fixed

**Pull request merged**